Handle race conditions caused by deleting a message and editing it.

abhijeetbodas2001 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes #16502 

**Issue**
Anything which updates a message can race with deleting it, leading to `DatabaseError`s. 
Messages are updated like-
- Fetch the Message and/or UserMessage with the `zerver/lib/message.py/access_message` function
- Update the instance
- Save the changes using `.save()` or `.update()`

If another thread deletes the Message after point 1 and before point 3, it leads to either `DatabaseError`s (when the Message itself was being saved) or `IntegrityError` (when some related field was being saved).

**Fix**
Both of these errors can be avoided by using `SELECT FOR UPDATE` database queries whenever the Message or related model is updated, which will lock the involved database rows untill the transaction is complete, and the message will be deleted after that.

**Testing**
The race conditions described above can be reproduced on the development server by opening two tabs- one normal and one private/incognito. Login with two different users, making sure one of them has rights to delete messages.
Add latency/delay in the message editing codepath, by using `time.sleep(5)` which will "pause" the request for 5 seconds. In those 5 seconds, delete the message from the other tab.




<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

abhijeetbodas2001 commented:
@mateuszmandera Can you review this?

abhijeetbodas2001 commented:
I'm not sure clear about these things-
- We already catch `IntegrityError` s in `do_add_reaction`, but for different reason (adding a reaction twice). Should I change the JsonableError message there? Os is there any way to classify IntegrityError? (I searched a bit, and could not find any way independent of the database)
- Is there any particular reason why we don't mock `Model.save` in tests? I couldn't any example in the tests. The `.save()` or `.update()` errors when caught had comments like `# nocoverage` beside them.
- I was thinking of catching `DatabaseError` [here](https://github.com/zulip/zulip/blob/master/zerver/lib/push_notifications.py#L800). Would it be necessary, given we are using `select_for_update` now onwards?

mateuszmandera commented on zerver/views/message_edit.py:
```diff
@@ -203,13 +205,20 @@ def update_message_backend(request: HttpRequest, user_profile: UserMessage,
+        # DatabaseError("Save with update_fields did not affect any rows.") occurs
+        # when the message is deleted in a race between calling access_message
+        # and saving the message.
+        # We use select_for_update to prevent this, but catch the error nonetheless.
```
I don't understand this part, if we're doing SELECT FOR UPDATE to prevent this, why are we catching? Isn't it supposed to be an impossible case? (and if somehow it's possible, then we missed something here and thus do want the exception to propagate to detect the bug)

abhijeetbodas2001 replied:
> Isn't it supposed to be an impossible case?
Yes, it will be impossible. Catching shouldn't be required. Should I remove it?
I looked at a few other PRs which made changes (in the frontend) so that exceptions would not be thrown, and they still caught the exceptions. I thought it was a part of good programming practice 😅

mateuszmandera replied:
Interesting, could you link some examples of those commits?

abhijeetbodas2001 replied:
#11351 is what I came across.

mateuszmandera replied:
Ah, those weren't really removing the possibility of exceptions occuring, so they did have to handled. I think we should remove this block here - it's not supposed to be possible to get there after all.

abhijeetbodas2001 replied:
Ok, makes sense. I'll remove that. Thanks!

mateuszmandera commented on zerver/views/submessage.py:
```diff
@@ -27,11 +29,14 @@ def process_submessage(request: HttpRequest,
+            content=content,
+        )
+    except IntegrityError:
+        raise JsonableError(_("Parent message was deleted."))
```
I'm wondering here whether we should be doing the `IntegrityError` catch or use `access_message` with `lock_message=True`. What was your reasoning for choosing to catch?

abhijeetbodas2001 replied:
Message row doesn't get updated in this codepath, so I thought it would be better performance wise to not lock it.

mateuszmandera replied:
Yeah, that's what's making me wonder too. Perhaps let's leave it as you've done it and wait for Tim or other reviewers.

timabbott replied:
I think it's correct to lock the message here -- if we think of the SubMessage as effectively a model containing extra fields for the Message (just like Reaction), `lock_message=True`  seems like the right semantics.  Certainly any change to SubMessage is going to end up flushing the Message from all caches, etc.

abhijeetbodas2001 commented:
@mateuszmandera Thanks a lot for reviewing this 😄

mateuszmandera commented:
>     * We already catch `IntegrityError` s in `do_add_reaction`, but for a different reason (adding a reaction twice). Should I change the JsonableError message there? Or is there any way to classify IntegrityError? (I searched a bit, but could not find any database-independent way to do so)

Yeah, that's problem with IntegrityErrors, we don't know what their cause is. Perhaps we can add another commit that'll SELECT FOR UPDATE the message before adding a reaction to it? It's ugly, but not sure what better options we have. Since that's something we might end up doing in various other places, we could put that logic in a helper function `ensure_message_exists_and_lock` or something like that?

>     * Is there any particular reason why we don't mock `Model.save` in tests? I couldn't find any example in the test suite. The `.save()` or `.update()` errors when caught had comments like `# nocoverage` beside them.

Where would we want/need to mock the `save` method?

>     * I was thinking of catching `DatabaseError` [here](https://github.com/zulip/zulip/blob/master/zerver/lib/push_notifications.py#L800). Would it be necessary, given we are using `select_for_update` now onwards?

Just like with the `message_edit` case discussed above, I don't think we should be catching it, as it's not supposed to happen anymore.

abhijeetbodas2001 commented:
> Yeah, that's problem with IntegrityErrors, we don't know what their cause is. Perhaps we can add another commit that'll SELECT FOR UPDATE the message before adding a reaction to it? It's ugly, but not sure what better options we have.

I did not understand why it would be ugly. Wouldn't changing the `lock_message` to true in the `access_message` call in views/reactions.py/add_reaction` suffice?

> Where would we want/need to mock the save method?

It was related to the third point only, so its resolved now 🙂

mateuszmandera commented:
> > Yeah, that's problem with IntegrityErrors, we don't know what their cause is. Perhaps we can add another commit that'll SELECT FOR UPDATE the message before adding a reaction to it? It's ugly, but not sure what better options we have.
> 
> I did not understand why it would be ugly. Wouldn't changing the `lock_message` to true in the `access_message` call in `views/reactions.py/add_reaction` suffice? Also, if we lock the message, I think we could also get remove the `IntegrityError` catch in `do_add_reaction`, since the race described in the comment there would not be possible right?


Yes, I think that could make sense. My one concern is performance - the earlier we acquire the lock, the more problematic this might become. I wonder if there's potential to end up with bad performance in a situation where a lot of people are posting reactions to the same message at once (for example because an announcement was posted and pinged everyone). But I think it's reasonable to add this commit and see if we want to merge it and give it a shot.

abhijeetbodas2001 commented:
> But I think it's reasonable to add this commit and see if we want to merge it and give it a shot.

> I think with that approach, we'd want to add a clear docstring in do_add_reactions that it's supposed to be called while holding the lock on the target message. We also have this snippet in onboarding.py where we'll probably want to select_for_update:

 I'm a bit busy with exams right now. I'll do this ASAP!

abhijeetbodas2001 commented:
I updated the PR after removing the `DatabaseError` catch as discussed.

> I think with that approach, we'd want to add a clear docstring in do_add_reactions that it's supposed to be called while holding the lock on the target message.

There are 12 calls to `do_add_reaction` (7 of them in tests). It wouldn't be required in tests or management commands right? I'm not sure about `zerver/openapi/curl_param_value_generators.py`


<details><summary>Grep output</summary>
<p>

```python
zilencer/management/commands/add_mock_conversation.py:        do_add_reaction(starr, preview_message, 'whale', emoji_code, reaction_type)
zilencer/management/commands/add_mock_conversation.py:        do_add_reaction(starr, preview_message, 'thumbs_up', emoji_code, reaction_type)
zerver/views/reactions.py:    do_add_reaction(user_profile, message, emoji_name, emoji_code, reaction_type)
zerver/tests/test_events.py:            lambda: do_add_reaction(
zerver/tests/test_events.py:        do_add_reaction(self.user_profile, message, "tada", "1f389", "unicode_emoji")
zerver/tests/test_management_commands.py:        do_add_reaction(self.example_user("iago"), message, "outbox", "1f4e4",  Reaction.UNICODE_EMOJI)
zerver/tests/test_management_commands.py:        do_add_reaction(self.example_user("hamlet"), message, "outbox", "1f4e4",  Reaction.UNICODE_EMOJI)
zerver/tests/test_management_commands.py:        do_add_reaction(self.mit_user("sipbtest"), message, "outbox", "1f4e4",  Reaction.UNICODE_EMOJI)
zerver/tests/test_import_export.py:        do_add_reaction(self.example_user("iago"), message, "outbox", "1f4e4",  Reaction.UNICODE_EMOJI)
zerver/tests/test_import_export.py:        do_add_reaction(self.example_user("hamlet"), message, "outbox", "1f4e4",  Reaction.UNICODE_EMOJI)
zerver/lib/onboarding.py:    do_add_reaction(welcome_bot, turtle_message, 'turtle', emoji_code, reaction_type)
zerver/openapi/curl_param_value_generators.py:    do_add_reaction(user_profile, message, emoji_name, emoji_code, reaction_type)
```

</p>
</details>

abhijeetbodas2001 commented:
@mateuszmandera I updated the PR. Please review this 🙂

timabbott commented:
@abhijeetbodas2001 can you try to avoid the duplication in the first commit?
```
-        message = Message.objects.select_related().get(id=message_id)
+        if lock_message:
+            message = Message.objects.select_for_update().select_related().get(id=message_id)
+        else:
+            message = Message.objects.select_related().get(id=message_id)
```

This would be better written using e.g. 
```
base_query = Message.objects.select_related()
if lock_message:
     base_query = base_select.select_for_update()
message = base_query.get(id=message_id)
```

so that we avoid duplicating the condition. (This is more important in the other call point where we're passing more parameters).

Also, is the thinking in that commit for why you pass `lock_*=False` in all callers (vs. not passing those parameters at all) just to make sure we're explicit everywhere by making them required arguments?  I think that's reasonable but we should explain that in the commit message.

timabbott commented:
I agree with @mateuszmandera's thoughts on docstrings in https://github.com/zulip/zulip/pull/16721#issuecomment-726022708 -- we should be explicit about the API.  I'm not worried about performance/races in this code path -- locking the single message that we're in the process of updating is very cheap.

abhijeetbodas2001 commented:
@timabbott Thanks for reviewing this! I updated the PR. Please take another look.
I wasn't sure how elaborate the docstrings need to be, so I just added one-liners saying "Message should be locked before calling this" in functions which are called after `access_message`. Should I describe the error which occurs and why it occurs more?

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -1866,16 +1866,14 @@ def notify_reaction_update(user_profile: UserProfile, message: Message,
 
 def do_add_reaction(user_profile: UserProfile, message: Message,
                     emoji_name: str, emoji_code: str, reaction_type: str) -> None:
+    """Should be called while holding a lock on the Message row, to prevent
+    IntegrityErrors.
+    """
```
I'd be more explicit here about the lock (since there are lots of various locks) and the things we're preventing with it, e.g.
```
Should be called while holding a FOR UPDATE lock on the Message row, to prevent double-adding the same reaction or race conditions with the message getting deleted concurrently.
```

timabbott replied:
We can mention (e.g. via `access_message(..., lock_message=True)` in the docstring as well.

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -774,8 +776,8 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
-                                                 lock_message=False,
-                                                 lock_usermessage=False)
+                                                 lock_message=True,
+                                                 lock_usermessage=True)
```
Is `lock_usermessage` sufficient?  I would expect so, because that's the only model we end up modifying, and the `Message` can't be deleted without deleted associated `UserMessage` rows.

timabbott replied:
I verified via manual `manage.py shell` testing that my theory here was correct.

timabbott commented:
I did a careful look at this, adjusted some comments, and pushed back here.  I think there's 3 major things I'd like to do before we merge this:
* I'd like to make sure that the RabbitMQ work happens outside the database transaction, to ensure that we don't hold these locks for a very long time if RabbitMQ is down and end up having a RabbitMQ outage result in a database outage as well. 
 I believe we can do this by just passing the `send_event` call via this Django feature: https://docs.djangoproject.com/en/3.0/topics/db/transactions/#django.db.transaction.on_commit.  Can you convert the submessage/reactions ones?  I think for `push_notifications`, we'd be better served by just doing a `with transaction.atomic()` around the two blocks starting with `access_message` and ending just before `message.trigger` gets set.
* I don't think the `update_message` code path is correct, because it already uses `@transaction.atomic` on do_update_message itself.  Let's defer that more complex case until we've done the rest to our satisfaction (maybe just drop the commit).

I think with those changes we can merge the reactions/submessage/push_notifications part of this PR, and then refocus on just the message editing code path.

As a sidenote, @andersk is there an easy way to make semgrep require that `access_message(lock_message=True)` is passed always inside an `@transaction.atomic()`?  I also kinda want a check for doing `transaction.atomic()` nested, since we generally don't do that intentionally.

abhijeetbodas2001 commented:
My end-semester exams start tomorrow so won't be able to do this immediately. I'll work on this as soon as I'm free.

mateuszmandera commented:
> I don't think the update_message code path is correct, because it already uses @transaction.atomic on do_update_message itself. Let's defer that more complex case until we've done the rest to our satisfaction (maybe just drop the commit).

@timabbott What's the concern with having nested `@transaction.atomic`? Doing `@transaction.atomic(savepoint=False)` on the (potentiallyy) nested transaction might also be the right fix.

abhijeetbodas2001 commented:
Updated the PR. Some tests were failing after doing `transaction.on_commit( lambda: send_event: ... )` because [transactions aren't commited in `TestCase`](https://docs.djangoproject.com/en/3.1/topics/db/transactions/#use-in-tests).
I mocked the `transaction.on_commit`. This is kind of hacky, and leads to repetition too. I found another way to do this as described [here](https://code.djangoproject.com/ticket/30457#comment:1). It defines a new context manager in `TestCase`. Would that be a better option for us?

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -1863,18 +1873,24 @@ def notify_reaction_update(user_profile: UserProfile, message: Message,
+
+    # We send the event after transaction is complete to make sure
+    # that an outage in RabbitMQ does not result in an outage in the
+    # database too.
```
It's not exactly about a " database outage" - the database will be fine, but specific rows would end up being locked FOR UPDATE for absurdly long, causing certain requests to freeze etc. if they tried to acquire a FOR UPDATE (or stronger) lock on the same row.

abhijeetbodas2001 replied:
Makes sense. I will change that (and the commit message too).
I was also thinking if we need these comments at all? Or would be sufficient to mention this in the commit message?

mateuszmandera replied:
Given that this is in multiple places, I'd favor putting it in the commit message rather than copypasting the same comment. That should be our general policy - "don't send to rabbitmq while holding database locks, use transaction.on_commit'" - so this seems like something that should actually go somewhere in the documentation (after being made more precise and better phrased) - not sure where though? :thinking: Perhaps @timabbott will have more thoughts here once he gets to this PR.

abhijeetbodas2001 replied:
Yeah, even I was skeptical about writing the same comment everywhere, but it didn't seem trivial to use `on_commit` to me so I wrote it anyways. It'd be great if this were written in the documentation. Maybe we could also enforce this in linters?

timabbott replied:
Yeah, I think `transaction.on_commit()` mostly speaks for itself, especially since we'll likely end up converting most `actions.py` `send_event` callers to use that pattern.

mateuszmandera commented:
@abhijeetbodas2001 Have you looked into whether it would be feasible to use `TransactionTestCase` for the base class in these situations? I imagine it might not be a particularly viable solution - in that case it seems feasible to do the `.on_commit` mocks - definitely seems better to have a nicely named wrapper like you've mentioned rather than manually `mock.patch`-ing. https://github.com/django/django/commit/e906ff6fca291fc0bfa0d52f05817ee9dae0335d adds another mechanism of dealing with the `.on_commit` problem, so we could a TODO in the wrapper to migrate to using that once we're on Django 3.2.

abhijeetbodas2001 commented:
> Have you looked into whether it would be feasible to use TransactionTestCase for the base class in these situations?

That would involve moving the relevant tests into a different class, which will extend `TransactionTestCase` right? Or make a new class similar to `ZulipTestCase` (which will extend `TransactionTestCase`) and extend that? I'm not sure. I think it would be difficult to keep these tests organized in that case, since they are in various parts of the test suite. 
Moreover, `TransactionTestCase` is known to be slow, since it clears the whole database after the test.
I think making a contextmanager would be a better idea, but please correct me if I'm wrong.

mateuszmandera commented:
Yeah makes sense, we can try the wrapper approach for now.

Btw, do you want to look into whether this PR fixes #16813 already and if not add commits for those races? I assume at least message starring might require some adjustment since we haven't touched that codepath here

abhijeetbodas2001 commented:
> Btw, do you want to look into whether this PR fixes #16813 already and if not add commits for those races? I assume at least message starring might require some adjustment since we haven't touched that codepath here

I'll investigate that and add commits to fix that here itself if that is fine 🙂

abhijeetbodas2001 commented:
@mateuszmandera Sorry for the delay. I updated the PR to use a wrapper, as discussed.
For #16813, I think using `UserMessage.objects.update_or_create` in both the "reactions" and "star" code, instead of `UserMessage.objects.create` should suffice?

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -1870,18 +1870,21 @@ def notify_reaction_update(user_profile: UserProfile, message: Message,
 
 def do_add_reaction(user_profile: UserProfile, message: Message,
                     emoji_name: str, emoji_code: str, reaction_type: str) -> None:
+    """Should be called while holding a select_for_update lock
+    (e.g. via access_message(..., lock_message=True)) on the
+    Message row, to prevent IntegrityErrors if the Message is deleted
+    by another thread while this is running.
```
I'd still phrase this more generally e.g. `.... to prevent race conditions.`. 

"if the Message is deleted by another thread while this is running." is too specific, this isn't the only case we're defending against, e.g. the lock should protect from double-adding a reaction and we're removing the block that was handling that case due to it.

As a sidenote, "thread" would also not be the best term here - we don't really use multithreading much - these races generally happen between two parallel processes, not threads - that's because parallel http requests get handled by the server in parallel processes.

abhijeetbodas2001 replied:
How about this?
```
Should be called while holding a FOR UPDATE lock
(e.g. via access_message(..., lock_message=True)) on the
Message row, to prevent IntegrityErrors due to race conditions.

```

mateuszmandera commented on zerver/views/reactions.py:
```diff
@@ -22,14 +23,17 @@ def create_historical_message(user_profile: UserProfile, message: Message) -> No
                                flags=UserMessage.flags.historical | UserMessage.flags.read)
 
+# transaction.atomic is required for the use of select_for_update in
+# access_message when lock_message or lock_usermessage is passed.
```
If we want these comments, they should be a bit more concise imo - they can just say that we run the function in a transaction to be a able to hold a FOR UPDATE lock throughout its course. (no need to elaborate on the implementation details of how `access_message` is called and what its arguments are called.

abhijeetbodas2001 replied:
Is this fine?
```
# transaction.atomic is required since we use FOR UPDATE queries in access_message
```

mateuszmandera commented on zerver/lib/actions.py:
```diff
@@ -1870,18 +1870,21 @@ def notify_reaction_update(user_profile: UserProfile, message: Message,
 
 def do_add_reaction(user_profile: UserProfile, message: Message,
                     emoji_name: str, emoji_code: str, reaction_type: str) -> None:
+    """Should be called while holding a select_for_update lock
```
Not sure if that's very update, but I'd rather name the lock by its proper name here - `FOR UPDATE`

mateuszmandera commented:
> For #16813, I think using `UserMessage.objects.update_or_create` in both the "reactions" and "star" code, instead of `UserMessage.objects.create` should suffice?

Wouldn't this be better solved by proper use of locks again? E.g. isn't there another race in the star codepath if the message is deleted while getting starred? Meaning we should address everything properly and only locks can guarantee that.

mateuszmandera commented:
> On a sidenote, if I do this
> 
>     * With `Zoe` logged in, view a stream which `Zoe` is not subscribed to (where the "You are not subscribed to stream ___" message is shown at the bottom along with the subscribe button)
> 
>     * Send a message to that stream from `Aaron`'s account, when `Aaron` is subscribed to that stream, from another instance.
> 
> 
> Then, the new message does not appear for `Zoe`, unless the tab is refreshed.
> Is this a bug or a performance optimization?

I think this is just the case of our event system not having the necessary pieces to let clients get events about a stream they're not subscribed to. Perhaps @timabbott will confirm.

abhijeetbodas2001 commented:
> Wouldn't this be better solved by proper use of locks again? E.g. isn't there another race in the star codepath if the message is deleted while getting starred? Meaning we should address everything properly and only locks can guarantee that.

I'm not really sure what kind of lock could be used here. Could you elaborate?
For the other race, we can do something similar to previous commits, using `FOR UPDATE` in this query in `actions.py/do_update_message_flags`

```py
msgs = UserMessage.objects.filter(user_profile=user_profile,
                                      message__id__in=messages)
```

mateuszmandera commented:
@abhijeetbodas2001 I was just thinking about locking the Message row. Would there be a problem with that?

abhijeetbodas2001 commented:
> @abhijeetbodas2001 I was just thinking about locking the Message row. Would there be a problem with that?

Got it, sorry for the confusion. Updated the PR with the changes in comments and a new commit.

timabbott commented:
> I think this is just the case of our event system not having the necessary pieces to let clients get events about a stream they're not subscribed to (https://github.com/zulip/zulip/pull/16721#issuecomment-746867977)

Correct, this is expected.  We may in the future change this, but doing so would be somewhat technically tricky.  (Maybe more relevant is we should just adjust the bottom-of-feed text to clarify that's what's happening).

timabbott commented:
For `test_events.py`, I think this approach better fits my vision for how this code should read (basically, avoiding changing every individual test by instead having the test class take care of this detail):

``` python
index 1419bbb42a..d89bfeff6e 100644
--- a/zerver/tests/test_events.py
+++ b/zerver/tests/test_events.py
@@ -246,7 +246,8 @@ class BaseAction(ZulipTestCase):
             include_streams=include_streams,
             realm=self.user_profile.realm,
         )
-        action()
+        with immediate_on_commit():
+            action()
         events = client.event_queue.contents()
         content = {
             'queue_id': '123.12',
@@ -556,17 +557,16 @@ class NormalActionsTest(BaseAction):
             stream_name=stream_name,
         )
 
-        with immediate_on_commit():
-            events = self.verify_action(
-                lambda: do_add_submessage(
-                    realm=cordelia.realm,
-                    sender_id=cordelia.id,
-                    message_id=message_id,
-                    msg_type='whatever',
-                    content='"stuff"',
-                ),
-                state_change_expected=False,
-            )
+        events = self.verify_action(
+            lambda: do_add_submessage(
+                realm=cordelia.realm,
+                sender_id=cordelia.id,
+                message_id=message_id,
+                msg_type='whatever',
+                content='"stuff"',
+            ),
+            state_change_expected=False,
+        )
 
         check_submessage('events[0]', events[0])
 
```

(We'll still need to manually use `immediate_on_commit` in other places that want the `send_event` calls to happen, though perhaps we can do it systematically in test helper methods like `tornado_redirected_to_list`?  Might need to migrate that function in a prep commit from being a `test-helpers.py` method to a feature of `ZulipTestCase`, which is probably a good chance anyway).

Also, can we replace `immediate_on_commit` with just backporting the Django method to ZulipTestCase, with a comment that we can delete it when we upgrade to Django 3.2?  It doesn't do anything when not called, and that avoids our using a non-standard API interface that we'll have to migrate later.

I'm still not altogether happy with this testing system plan even with those changes, because any test where we neglect to add this logic that is testing a place we're using `transaction.on_commit` will simply never run the `on_commit()` code, which will substantially reduce coverage.  Maybe we can do some sort of fancy linting approach to catch this sort of situation?  @andersk @showell FYI; maybe we should do a chat.zulip.org thread explaining the full context.

What I'd prefer is for, in our tests, the `on_commit` hooks to just be run immediately when we exit the nearest `transaction.atomic()` block (and maybe error in some nesting casing), but it sounds like Django doesn't (yet) have a good way to do that.

abhijeetbodas2001 commented:
>Also, can we replace immediate_on_commit with just backporting the Django method to ZulipTestCase, with a comment that we can delete it when we upgrade to Django 3.2?

Yes. Shall I go ahead with the approach you described?

> What I'd prefer is for, in our tests, the on_commit hooks to just be run immediately when we exit the nearest transaction.atomic() block (and maybe error in some nesting casing),

@timabbott That could be achieved by wrapping the `TestCase.run()` function itself in `immediate_on_commit` right? In any case, I think something like this is still much DRYier than the PR now.

```python
--- a/zerver/lib/test_classes.py
+++ b/zerver/lib/test_classes.py
@@ -49,7 +49,7 @@ from zerver.lib.test_console_output import (
     TeeStderrAndFindExtraConsoleOutput,
     TeeStdoutAndFindExtraConsoleOutput,
 )
-from zerver.lib.test_helpers import find_key_by_email, instrument_url
+from zerver.lib.test_helpers import find_key_by_email, instrument_url, immediate_on_commit
 from zerver.lib.users import get_api_key
 from zerver.lib.validator import check_string
 from zerver.lib.webhooks.common import get_fixture_http_headers, standardize_headers
@@ -123,6 +123,7 @@ class ZulipTestCase(TestCase):
                 self.mock_ldap.reset()
             self.mock_initialize.stop()

+    @immediate_on_commit()
     def run(self, result: Optional[TestResult]=None) -> Optional[TestResult]:  # nocoverage
         if not settings.BAN_CONSOLE_OUTPUT:
             return super().run(result)
```

mateuszmandera commented:
@abhijeetbodas2001 There's also a race in `remove_reaction` codepaht (e.g. two concurrent requests trying to remove the same reaction. Also, I think, a request removing a reaction concurrently with a request deleting the message). My idea here would be to do the standard thing and select for update the Message. It'd require using `.on_commit` on `notify_reaction_update` so that change should go in this PR. Do you have time to add that commit?

abhijeetbodas2001 commented:
@mateuszmandera Sure. Added one more commit.

Also, I removed the `immediate_on_commit` code-blocks. Tests are failing as expected.
Waiting for comments on [16721#issuecomment-749319756](https://github.com/zulip/zulip/pull/16721#issuecomment-749319756) 🙂

abhijeetbodas2001 commented:
@timabbott Could you take a look at my [comment above](https://github.com/zulip/zulip/pull/16721#issuecomment-749319756)?

timabbott commented:
@abhijeetbodas2001 it'd be great if you can do a rebase on this, I think across the `black` transition.  

I think we don't want to do that hacky approach with `immediate_on_commit` around every test, just because that'll make it impossible to test some of the race situations?  Not sure.

abhijeetbodas2001 commented:
@timabbott Rebased, along with a few changes-
- Cleaned up the commit messages a bit to not echo the diff :P
- `immediate_on_commit` is now a method of `ZulipTestCase`, not a standalone helper
- Wrapped the `action()` call in `test_events.py/verify_action()` with `immediate_on_commit`, as you mention in the comment above.
- In a few other places which test reactions/submessage events, manually wrapper the API calls with `immediate_on_commit`. I'd appreciate it if you could elaborate a bit on what you mean by utilizing `tornado_redirected_to_list` 😅

timabbott commented:
Marking this as something to pick up after the 4.0 release.  This is really important work of our transaction model but also has some technical risk.

mateuszmandera commented:
@abhijeetbodas2001 Just a heads up that with Django 3.2 upgrade merged, we now have access to `captureOnCommitCallbacks` that iirc was discussed here as something that would let us get around some hacks that were needed in tests

timabbott commented:
Great, yeah, let's definitely rework this to fit that model.

abhijeetbodas2001 commented:
Sure, will open a new PR.

abhijeetbodas2001 commented:
Closing in favor of #18559
